File: typescript/src/exec.ts
import { PingClient } from "./client.js";
import { StatsLogger } from "./logging.js";
import { ResponseHandler } from "./response.js";
import { PingStats } from "./stats.js";

export class Executor {
  constructor(
    private client: PingClient,
    private handler: ResponseHandler,
    private logger: StatsLogger
  ) {}

  async execute(url: string, nbTries: number) {
    const stats = new PingStats();

    for (let i = 0; i < nbTries; i++) {
      const response = await this.client.ping(url);

      this.handler.updateStats(stats, response);
      this.logger.showCurrentStats(stats);
    }

    this.logger.showStatsSummary(stats);
  }
}

File: rust/src/exec.rs
use std::sync::Arc;

use crate::{
    client::PingClient, logging::StatsLogger, response::ResponseHandler, stats::PingStats,
};

pub struct Executor {
    client: Arc<dyn PingClient>,
    handler: Arc<ResponseHandler>,
    logger: Arc<dyn StatsLogger>,
}

impl Executor {
    pub fn new(
        client: Arc<dyn PingClient>,
        handler: Arc<ResponseHandler>,
        logger: Arc<dyn StatsLogger>,
    ) -> Self {
        Self {
            client,
            handler,
            logger,
        }
    }

    pub fn execute(&self, url: &str, nb_tries: u32) {
        let mut stats = PingStats::new();

        for _ in 0..nb_tries {
            let response = self.client.ping(url);

            self.handler.update_stats(&mut stats, &response);
            self.logger.show_current_stats(&stats);
        }

        self.logger.show_stats_summary(&stats);
    }
}
